### PR TITLE
JVM IR: Generate synthetic methods for type aliases with annotations

### PR DESCRIPTION
**Pull request by kandersen**

### CODE REVIEW

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/TypeAliasAnnotationMethodsLowering.kt [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        declaration.acceptChildrenVoid(this)
+        declaration.visitTypeAliases()
+    }
+
+    private val IrTypeAlias.syntheticAnnotationMethodName
+        get() = Name.identifier(JvmAbi.getSyntheticMethodNameForAnnotatedTypeAlias(name))
+
+    private fun IrDeclarationContainer.visitTypeAliases() {
+        val annotatedAliases = declarations.filterIsInstance<IrTypeAlias>().filter { !it.descriptor.annotations.isEmpty() }
+        for( alias in annotatedAliases ) {
```
Comment (max-kammerer) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/TypeAliasAnnotationMethodsLowering.kt]: Please reformat

File: compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/declarations/IrDeclarationOrigin.kt [hunk 2]
```diff
@@ -42,6 +42,8 @@ interface IrDeclarationOrigin {
     object IR_BUILTINS_STUB : IrDeclarationOriginImpl("IR_BUILTINS_STUB")
     object BRIDGE : IrDeclarationOriginImpl("BRIDGE", isSynthetic = true)
     object BRIDGE_SPECIAL: IrDeclarationOriginImpl("BRIDGE_SPECIAL")
+    object SYNTHETIC_METHOD_FOR_TYPEALIAS_ANNOTATIONS :
```
Comment (max-kammerer) [inline compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/declarations/IrDeclarationOrigin.kt]: As it's only jvm-specific please move it to JvmLoweredDeclarationOrigin

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/TypeAliasAnnotationMethodsLowering.kt [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+    override fun lower(irFile: IrFile) {
+        irFile.acceptVoid(this)
+    }
+
+    override fun visitElement(element: IrElement) {
+        element.acceptChildrenVoid(this)
+    }
+
+    override fun visitFile(declaration: IrFile) {
```
Comment (max-kammerer) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/TypeAliasAnnotationMethodsLowering.kt]: This override is redundant cause of fileClassPhase, that moves file content into class

Comment (max-kammerer) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/TypeAliasAnnotationMethodsLowering.kt]: ...and lower could be switched to ClassLoweringPass

Comment (kandersen) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/TypeAliasAnnotationMethodsLowering.kt]: Is the same true of PropertiesLowering?

Comment (kandersen) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/TypeAliasAnnotationMethodsLowering.kt]: It is! Just tried tried out the refactoring of PropertiesLowering to a ClassLoweringPass. Do you want that in this PR, or should I extract that to a separate one?

Comment (max-kammerer) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/TypeAliasAnnotationMethodsLowering.kt]: It's best to make separate ibe

Comment (max-kammerer) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/TypeAliasAnnotationMethodsLowering.kt]: Oh, no. PropertiesLowering is common lowering. And as i remeber Native and JS don't perform fileClassPhase

Comment (kandersen) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/TypeAliasAnnotationMethodsLowering.kt]: Ah, right! I just spotted the same as I went to check the changes in and I noticed the path was different.

File: compiler/ir/backend.common/src/org/jetbrains/kotlin/ir/builders/declarations/declarationBuilders.kt [hunk 4]
```diff
@@ -157,7 +157,7 @@ inline fun buildFun(builder: IrFunctionBuilder.() -> Unit): IrFunctionImpl =
         buildFun()
     }
 
-inline fun IrDeclarationContainer.addFunction(builder: IrFunctionBuilder.() -> Unit): IrSimpleFunction =
+inline fun IrDeclarationContainer.addFunction(builder: IrFunctionBuilder.() -> Unit): IrFunctionImpl =
```
Comment (max-kammerer) [inline compiler/ir/backend.common/src/org/jetbrains/kotlin/ir/builders/declarations/declarationBuilders.kt]: Is this change is neccessery for pr?

Comment (kandersen) [reply compiler/ir/backend.common/src/org/jetbrains/kotlin/ir/builders/declarations/declarationBuilders.kt]: Yes, it allows line 54 in TypeAliasAnnotationMethodsLowering.kt. It's, as far as I can tell, an unnecessary restriction to the superclass.

**Comment by max-kammerer** [comment]
Many thanks!

**Pull request merged**